### DISCUSSION TITLE
Rendering to Table without a key prop

### DISCUSSION THREAD

**Discussion by binary64** [ant-design/ant-design#41932] [Q&A]
Hi everyone, I am using Apollo Client's React Hooks to fetch data:
```
const { data } = useSomeQuery()
return <Table data={data} />
```
The Table component then maps through the data array and renders the rows. But it is missing the 'key' prop. What are my options?

My first thought is to do `data={data.map(e => ({ ...e, key: e.id }))}` but this would cause a big drop in React performance right?

How would you solve this problem, whilst keeping React performant

**Comment by abenhamdine** [comment]
>  but this would cause a big drop in React performance right?

Why ?

**Reply by binary64** [reply]
Because map() returns a new object every time it is rendered, and React will see this as a change to the DOM.

**Reply by abenhamdine** [reply]
You should first check if there's really a perf issue, because React is quite performant at rerendering the DOM (thanks to the virtual DOM).
But if it's the case, you could memoize the component with memo https://react.dev/reference/react/memo

**Reply by binary64** [reply]
abenhamdine: Should I check every single component for this? What is a good test setup for this? Should I set over multiple CPUs and mobile devices? Should I plot the timing on a chart to determine the coefficiencies? Will memoizing one component mean I should re-test the others?
